doc/guides/coding_guidelines/index.rst

```diff
@@ -784,3 +787,30 @@ and the equivlent rules from other standards for reference.
       -  Where designated initializers are used to initialize an array object the size of the array shall be specified explicitly
       -  N/A
       -  `Rule 9.5 <https://gitlab.com/MISRA/MISRA-C/MISRA-C-2012/Example-Suite/-/blob/master/R_09_05.c>`_
+
+Additional rules
+****************
+
+Rule A.1: Conditional Compilation
+=================================
+
+Severity
+--------
+
+Required
+
+Description
+-----------
+
+Do not conditionally compile function declarations in header files.  Do not
+conditionally compile structure declarations in header files.  You may
+conditionally exclude fields within structure definitions to avoid wasting
+memory when the feature they support is not enabled.
+
+Rationale
+---------
+
+Excluding declarations from the header based on compile-time options may prevent
+their documentation from being generated. Their absence also prevents use of
+``if (IS_ENABLED(CONFIG_FOO)) {}`` as an alternative to preprocessor
+conditionals when the code path should change based on the selected options.
```

Comment:
Shouldn't there be other exceptions listed? Look at many of the header files now, for example CONFIG_USERSPACE conditionals are needed in many of the places.

Reply:
Isn't this rule about conditionally compile functions and types in headers ? How this is related with `IS_ENABLED` that is used with generated symbols ?

Reply:
@ceolin If the material enabled by `CONFIG_DEVICE_POWER_MANAGEMENT` were unconditionally declared in `<device.h>` rather than wrapped in `#ifdef CONFIG_DEVICE_POWER_MANAGEMENT` you could do:

    int rc = -ENOTSUP;
    struct device *cons = device_get_binding(DT_LABEL(DT_CHOSEN(zephyr_console)));

    if (IS_ENABLED(CONFIG_DEVICE_POWER_MANAGEMENT)) {
        rc = device_set_power_state(cons, DEVICE_PM_ACTIVE_STATE, NULL, NULL);
    }

rather than having to do:

    int rc = -ENOTSUP;
    struct device *cons = device_get_binding(DT_LABEL(DT_CHOSEN(zephyr_console)));

    #ifdef CONFIG_DEVICE_POWER_MANAGEMENT
    rc = device_set_power_state(cons, DEVICE_PM_ACTIVE_STATE, NULL, NULL);
    #endif

because the latter generates build errors when `device_set_power_state` and `DEVICE_PM_ACTIVE_STATE` are undefined as they are when `CONFIG_DEVICE_POWER_MANAGEMENT=n` currently.

Some developers find the former to be easier to understand.

Reply:
if `CONFIG_DEVICE_POWER_MANAGEMENT` is undefined `device_set_power_state` won't be compile so no error should happen. Am I missing something ?

Regarding the macro `IS_ENABLED` it should no be encouraged, actually should be removed. It violates rule 14.3 `Controlling expressions shall not be invariant`.

Reply:
> if `CONFIG_DEVICE_POWER_MANAGEMENT` is undefined `device_set_power_state` won't be compile so no error should happen. Am I missing something ?

I think so, but I can't really tell what.  Somebody else might be able to explain it better.

> Regarding the macro `IS_ENABLED` it should no be encouraged, actually should be removed. It violates rule 14.3 `Controlling expressions shall not be invariant`.

Heh.  OK.  Should be interesting to see how people respond to that one.

Reply:
> Should be interesting to see how people respond to that one.

Here it comes :)

> Some developers find the former to be easier to understand.

`IS_ENABLED()` isn't just about readability. Using #ifdefs means you increase the number of build-time configurations both the developer and CI needs to test in order to guarantee that everything builds fine. With `IS_ENABLED()` the compiler will always "see" the code and let you know of syntax errors, etc, even if the Kconfig option in question is not enabled.

With #ifdefs and CI, in the best case (when the developer remembered to add build tests for all configuration combinations) you've increased the load and and the time CI takes to complete (which is already a problem for us). In the worst case you've introduced code to the tree that's never build tested and possibly broken.

> Regarding the macro IS_ENABLED it should no be encouraged, actually should be removed. It violates rule 14.3 Controlling expressions shall not be invariant.

I'd argue that both the improved readability aspect and the reduced number of builds & configurations to verify syntactical correctness of the code contribute to improved code quality, and therefore any coding guideline requiring the opposite would be self defeating.

Reply:
> It violates rule 14.3 Controlling expressions shall not be invariant.

Just wanted to add to this that I don't think using IS_ENABLED() with Kconfig symbols is necessarily covered by that rule (or at least its spirit), since we're not talking about something that's completely invariant here. We might want to still have this rule but make it clear that it doesn't apply to values which may change based on build-time configuration.

Reply:
> > Should be interesting to see how people respond to that one.
> 
> Here it comes :)
> 
> > Some developers find the former to be easier to understand.
> 
> `IS_ENABLED()` isn't just about readability. Using #ifdefs means you increase the number of build-time configurations both the developer and CI needs to test in order to guarantee that everything builds fine. With `IS_ENABLED()` the compiler will always "see" the code and let you know of syntax errors, etc, even if the Kconfig option in question is not enabled.

Ok, that I understand but that does not mean build errors when `CONFIG_DEVICE...` is undefined, that was the reason of my question :)

The unconditional syntax check is something that I can't argue against, it is a good thing.

> 
> With #ifdefs and CI, in the best case (when the developer remembered to add build tests for all configuration combinations) you've increased the load and and the time CI takes to complete (which is already a problem for us). In the worst case you've introduced code to the tree that's never build tested and possibly broken.
> 
> > Regarding the macro IS_ENABLED it should no be encouraged, actually should be removed. It violates rule 14.3 Controlling expressions shall not be invariant.
> 
> I'd argue that both the improved readability aspect and the reduced number of builds & configurations to verify syntactical correctness of the code contribute to improved code quality, and therefore any coding guideline requiring the opposite would be self defeating.

I don't completely agree with the readability argument, the unconditional syntax check is really something to consider. Regarding the rule itself we definitely can document this deviation or "reinterpretation". But keep in mind that this rule won't be the only one, this will cause other violations due unreachable code.